File: Python/ByRound/2067/2067_E_White_Magic.py
# Problem: CF 2067 E - White Magic
# https://codeforces.com/contest/2067/problem/E

"""
E. White Magic

Purpose:
This code finds the maximum possible length of a magical subsequence from a given sequence of non-negative integers.
A magical sequence satisfies the condition that for all prefixes, the minimum element is greater than or equal to the mex (minimum excluded value) of the suffix.

Algorithms/Techniques:
- Greedy approach with set-based MEX computation
- Reverse iteration through array to compute MEX efficiently

Time Complexity: O(n) per test case, where n is the length of input array.
Space Complexity: O(n) for storing elements in a set.

Input Format:
- First line contains number of test cases t.
- For each test case:
  - First line contains n (length of sequence)
  - Second line contains n integers (the sequence)

Output Format:
- For each test case, output the maximum possible length of a magical subsequence.
"""

import sys

input = sys.stdin.readline
for i in range(int(input())):
    n = int(input())
    a = list(map(int, input().split()))
    s = set()
    x = a.index(0) if 0 in a else -1  # Position of first zero, or -1 if no zero
    mex = 0
    for j in range(n - 1, -1, -1):  # Iterate from end to start
        if a[j] < mex:  # If current element is less than current mex, reset mex to 0 and break
            mex = 0
            break
        if a[j] or j == x:  # Add element to set if it's non-zero or it's the first zero
            s.add(a[j])
            while mex in s:  # Compute next mex by checking what's missing from the set
                mex += 1
    print(n - a.count(0) + int(mex != 0))  # Output result based on count of zeros and whether mex was updated


# https://github.com/VaHiX/codeForces/

File: CPP/ByTier/E/2067_E_White_Magic.cpp
// Problem: CF 2067 E - White Magic
// https://codeforces.com/contest/2067/problem/E

/*
E. White Magic
Algorithms/Techniques: Greedy with MEX calculation and tracking of prefix minimums.
Time Complexity: O(n) amortized per test case, due to the use of a mex counter and careful updates.
Space Complexity: O(n), for storing arrays of size up to 2*10^5.

The task is to find the maximum length of a magical subsequence in a given array,
where a sequence is magical if for all prefixes, the minimum of that prefix
is greater than or equal to the mex of the suffix starting after that prefix.

Approach:
- Process the array from right to left.
- Keep track of how many zeros are encountered (c).
- For each element:
  - If it's not zero or we have only one zero remaining, increment its count in cnt array.
  - Update the mex value by advancing while there are elements in the range [0, mex].
- Compute final answer using a formula involving the total count of zeros and modified tk.

*/

#include <iostream>

using namespace std;
#define int long long
int a[200005], cnt[200005];

void DO() {
  int n;
  cin >> n;
  int tk = n, c = 0, cc; // tk: total count of elements to consider. c: number of zeros seen so far. cc: original count of zeros.
  for (int i = 1; i <= n; i++) {
    cin >> a[i];
    c += a[i] == 0; // Count zeros
    cc = c;         // Save initial zero count
  }
  for (int i = n, mex = 0; i >= 1; i--) {
    if (a[i] < mex)
      tk--; // If current element is less than mex, exclude it from result
    
    if (a[i] <= n and (a[i] != 0 or c == 1))
      cnt[a[i]]++; // Only count elements in [0, n], unless they are zeros with only one left
    
    c -= a[i] == 0; // Decrease zero counter as we move back
    
    while (cnt[mex]) // Increment mex if current mex is filled
      mex++;
  }
  cout << n - cc + (tk == n and cc) << '\n'; 
  for (int i = 0; i <= n; i++)
    cnt[i] = 0; // Reset counter array
}

signed main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int t, i = 1;
  cin >> t;
  while (t--) {
    DO();
  }
}


// https://github.com/VaHiX/codeForces/